contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1089 @@
+   *      - if the operation is a revocation, that permission is enforced by the TOKEN_REVOKER_ROLE
+   *      - if the operation is a withdrawal the burn amount must be <= the sender's claimable balance
+   *    - the operation is a transfer and _all_ of the following must be true:
+   *      - the operator is operating on their own balance (enforced in the inherited contract)
+   *      - the operator has sufficient balance to transfer (enforced in the inherited contract)
+   */
+  function _beforeTokenTransfer(
+    address operator,
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  )
+    internal
+    override(ERC1155PresetMinterPauserUpgradeable, ERC1155SupplyUpgradeable)
+  {
+    bool isMinting = from == address(0);
+    bool isBurning = to == address(0);
+
+    bool isWithdrawing = isBurning && from == operator;
+    bool isTransferring = !isMinting && !isBurning;
+
+    if (isBurning) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        Schedule storage schedule = _scheduleIdToScheduleStruct[ids[i]];
+        schedule.releasedAmountFloor = _releasedBalanceOfSingleSchedule(ids[i]);
+      }
+    }
+    if (isWithdrawing) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        if (amounts[i] > claimableBalanceForScheduleForAccount(ids[i], from)) {
+          revert InsufficientClaimableBalance({
+            account: from,
+            scheduleId: ids[i]
+          });
+        }
+      }
+    }
+    return super._beforeTokenTransfer(operator, from, to, ids, amounts, data);
+  }
+
+  /**
+   * Linearly released balance for a single schedule at the current block timestamp, ignoring any
+   * released amount floor that has been set for the schedule.
+   */
+  function _linearReleaseAmountAvailable(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (block.timestamp >= schedule.endTime) {
+      return totalSupply(scheduleId);
+    }
+    uint256 rampTotalTime = schedule.endTime - schedule.startTime;
+    return
+      block.timestamp < schedule.startTime
+        ? 0
+        : (totalSupply(scheduleId) * (block.timestamp - schedule.startTime)) /
+          rampTotalTime;
+  }
+
+  /**
+   * @dev Calculates the number of tokens that should be revoked from a given token holder and schedule based on their
+   * proportion of ownership of that schedule's tokens and the total number of tokens being revoked.
+   */
+  function _quantityToRevokePerTokenHolder(
+    uint256 totalQuantityToRevoke,
+    uint256 scheduleId,
+    address account
+  ) private view returns (uint256) {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 scheduleTrueTotal = schedule.totalClaimedAmount +
+      totalSupply(scheduleId);
+    // avoid division by 0
+    if (scheduleTrueTotal == 0) {
+      return 0;
+    }
```

Comment:
could this be refactored so that this function only has a single return statement?